File: C#/ConsoleApp1/ConsoleApp1/Combination.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;

namespace ClassLibrary1
{
    public class Combination
    {
        public enum Operators
        {
            PLUS,
            MINUS,
            MULTIPLY,
            DIVIDE,
            POWER,
            ROOT
        };
        
        private List<Operators> operations = new List<Operators> {Operators.PLUS,Operators.MINUS,Operators.DIVIDE,Operators.MULTIPLY,Operators.POWER,Operators.ROOT};
        List<Solution> solutions = new List<Solution>();
        public List<Solution> find_best_move(List<string> input)
        {
            solutions = new List<Solution>();
            permutations(new List<string>(), input);
            solutions.Reverse();
            int j = 0;
            foreach (var solution in solutions)
            {
                Console.WriteLine(j++ +  ".    " + solution);
            }

            return solutions;
        }

        public void permutations(List<String> path, List<String> nodes)
        {
            foreach (var node in nodes)
            {
                if (path.Count(x => x == node) < nodes.Count(x => x == node))
                {
                    path.Add(node);
                    Solution solution = divide_to_strings(path,new List<String> {"","",""},0,0,true);
                    if (solution != null){
                        check_best(solution);
                    }
                    permutations(path, nodes);
                    path.Remove(node);
                }
            }
        }
        
        public Solution divide_to_strings(List<string> nodes,List<string> divides,int index,int level,bool jumper){
            
            /*foreach (var node in nodes)
            {
                Console.Write(node + " ");
            }
            Console.WriteLine();
            foreach (var node in divides)
            {
                Console.Write(node + " ");
            }
            Console.WriteLine();*/
            level += 1;
            if (divides.Any(s => s.Length > 0 && s[0] == '0')) return null;

            if (nodes.Count == 0) return null;
            List<string> nodesStart = new List<string>(nodes);
            List<string> dividesStart = new List<string>(divides);

            if (nodesStart.Count > 0){
                Solution jump = null;

                String node = nodesStart[0];
                nodesStart.RemoveAt(0);
                
                //preskoci
                if (jumper) jump = divide_to_strings(nodesStart,dividesStart,index,level++,true);

                String replacer = dividesStart[index];

                if (replacer == "" && node == "0") return null;

                replacer += node;
                dividesStart[index] = replacer;


                Solution solution = Control(dividesStart[0],dividesStart[1],dividesStart[2]);



                //bud ich da dokopy
                Solution unite = divide_to_strings(nodesStart,dividesStart,index,level++,false);
                //alebo rozdeli
                Solution divide = index + 1 < 3 ? divide_to_strings(nodesStart,dividesStart,index + 1 ,level++,false) : null ;
                Solution pom = ChooseParent(unite,divide,jump,solution);
                return pom;
            }
            return null;
        }
        
        public bool contains_all(Solution stonesFather,Solution stonesSon){
            //Console.Write(stonesFather + " ||| " + stonesSon);
            foreach (var i in stonesSon.stones)
            {
                if (stonesFather.stones.Count(x => x == i) < stonesSon.stones.Count(x => x == i))
                {
                    //Console.WriteLine(" FALSE");
                    return false;
                }
            }
            //Console.WriteLine(" TRUE");
            return true;
        }

        public Solution ChooseParent(Solution parent, Solution parent2, Solution parent3, Solution actual)
        {
            // ak ma tri nadpriklady
            Solution outer = null;
            List<Solution> parents = new List<Solution> {parent, parent2, parent3};
            if (parents.Count(x => x == null) == 0){
                if (parent2.value >= parent.value && parent2.value >= parent3.value){
                    parents.Remove(parent2);
                    outer = parent2;
                }
                else if (parent3.value >= parent.value){
                    parents.Remove(parent3);
                    outer = parent3;
                }
                else {
                    parents.Remove(parent);
                    outer = parent;
                }

                foreach (var s in parents)
                {
                    if (contains_all(outer,s)){
                        outer.add_child(s);
                    }
                }
            }
            //ak dvaja
            else if (parents.Count(x => x == null) == 1){
                parents.Remove(null);
                if (parents[0].value >= parents[1].value) {

                    outer = parents[0]; //odstrani lepsieho
                    parents.RemoveAt(0);
                }
                else {
                    outer = parents[1];
                    parents.RemoveAt(1);
                }
                foreach (var s in parents){
                    if (contains_all(outer,s)){
                        outer.add_child(s);
                    }
                }
            }
            //ak iba jeden
            else if (parents.Count(x => x == null) == 2){
                parents.RemoveAll(x => x == null);
                outer = parents[0];
            }
            //ak ziaden
            else if (actual != null){
                outer = actual;
            }

            if (actual != null && outer != null){
                if (contains_all(outer,actual)){
                    outer.add_child(actual);
                }
            }
            return outer;
        }

        private Solution Control(string one, string two, string three){

            int oneNum = ToInt(one);
            int twoNum = ToInt(two);
            int threeNum = ToInt(three);

            if (twoNum == -1 && threeNum == -1) return null;


            List<Solution> solutionValues;
            if (threeNum == -1){
                solutionValues = check_unary_solution(oneNum,twoNum);
            }
            else {
                solutionValues = check_binary_solution(oneNum,twoNum,threeNum);
            }
            if (solutionValues.Count > 0){
                Solution parent = solutionValues[0];
                solutionValues.RemoveAt(0);
                foreach (var s in solutionValues) parent.add_child(s);
                return parent;
            }
            return null;
        }
        
        private static int ToInt(String val) {
            return val == "" ? -1 : Convert.ToInt32(val);
        }
        
        public void check_best(Solution solution){
            bool isThere = true;
            foreach (var s in solutions){
                if (s.priklad == solution.priklad) isThere = false;
            }

            if (solutions.Count < 10 && isThere){
                solutions.Add(solution);
            }
            else if (solutions[0].value < solution.value && isThere){
                solutions.RemoveAt(0);
                solutions.Add(solution);
            }
            solutions.Sort();
        }

        private List<Solution> check_unary_solution(int oneNum,int twoNum){
            List<Solution> outer = new List<Solution>();
            if (check_operator_in_play(Operators.POWER) && check_power_of(oneNum,twoNum)){
                outer.Add(new Solution(count_digits(oneNum) + count_digits(twoNum),oneNum + " ** " + twoNum,make_stones_list(oneNum,twoNum,-1)));
            }
            if (check_operator_in_play(Operators.ROOT) && check_root(oneNum,twoNum)){
                outer.Add(new Solution(count_digits(oneNum) + count_digits(twoNum),oneNum + " /- " + twoNum,make_stones_list(oneNum,twoNum,-1)));
            }
            return outer;
        }

        private List<Solution> check_binary_solution(int oneNum,int twoNum,int threeNum){
            List<Solution> outer = new List<Solution>();
            if (check_operator_in_play(Operators.PLUS) && check_plus(oneNum,twoNum,threeNum)){
                outer.Add(new Solution(count_digits(oneNum) + count_digits(twoNum),oneNum + " + " + twoNum + " = " + threeNum,make_stones_list(oneNum,twoNum,threeNum)));
            }
            if (check_operator_in_play(Operators.MINUS) && check_minus(oneNum,twoNum,threeNum)){
                outer.Add(new Solution(count_digits(oneNum) + count_digits(twoNum),oneNum + " - " + twoNum + " = " + threeNum,make_stones_list(oneNum,twoNum,threeNum)));
            }
            if (check_operator_in_play(Operators.MULTIPLY) && check_multiply(oneNum,twoNum,threeNum)){
                outer.Add(new Solution(count_digits(oneNum) + count_digits(twoNum),oneNum + " * " + twoNum + " = " + threeNum,make_stones_list(oneNum,twoNum,threeNum)));
            }
            if (check_operator_in_play(Operators.DIVIDE) && check_divide(oneNum,twoNum,threeNum)){
                outer.Add(new Solution(count_digits(oneNum) + count_digits(twoNum),oneNum + " / " + twoNum + " = " + threeNum,make_stones_list(oneNum,twoNum,threeNum)));
            }
            return outer;
        }

        private int count_digits(int input){
            int output = 0;
            while (input > 0){
                output += input % 10;
                input /= 10;
            }
            return output;
        }

        private List<String> make_stones_list(int one,int two,int three){
            String s = one + two.ToString();
            if (three >= 0) s += three.ToString();
            List<String> outer = new List<String>();
            foreach (var c in s)
            {
                outer.Add(Convert.ToString(c));
            }
            return outer;
        }

        private bool check_operator_in_play(Operators o){
            return operations.Contains(o);
        }

        private bool check_power_of(int oneNum,int twoNum){
            return Math.Pow(oneNum,2) == twoNum || Math.Pow(oneNum,3) == twoNum;
        }
        public bool check_root(int one_num,int two_num){
            return Math.Sqrt(one_num) == two_num || Math.Pow(one_num, (double) 1 / 3) == two_num;
        }
        public bool check_plus(int one_num,int two_num,int three_num){
            return one_num + two_num == three_num;
        }
        public bool check_minus(int one_num,int two_num,int three_num){
            return one_num - two_num == three_num;
        }
        public bool check_multiply(int one_num,int two_num,int three_num){
            return one_num * two_num == three_num;
        }
        public bool check_divide(int one_num,int two_num,int three_num){
            return (double)one_num / two_num == three_num;
        }
    }
    
}

File: Java/src/Combination.java
import javafx.util.Pair;

import javax.xml.soap.Node;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Combination {
    private int max = 0;
    private List<Operators> operations = Arrays.asList(Operators.values());
    private List<Solution> solutions = new ArrayList<>();

    /*public static void main(String[] args) {
        find_best_move("1 2 3 4 5 6 7 8 2 5");
    }*/

    /**
     *
     * @param input
     */
    public List<Solution> find_best_move(List<String> input){
        solutions = new ArrayList<>();
        permutations(new ArrayList<>(),input);
        System.out.println("======================================");
        //System.out.println(solutions);
        Collections.reverse(solutions);
        int i = 0;
        for (Solution s : solutions) System.out.println(i++ + ".     " + s);
        System.out.println(input);
        return solutions;
    }

    public void permutations(List<String> path,List<String> nodes){
        //dostane vstup a vyrobi vsetky mozne permutacie kamenov

        for (String node: nodes){
            if (Collections.frequency(path, node) < Collections.frequency(nodes, node)) {
                path.add(node);
                Solution solution = divide_to_strings(path,new ArrayList<>(Arrays.asList("","","")),0,0,true);
                if (solution != null){
                    check_best(solution);
                }


                permutations(path, nodes);
                path.remove(node);
            }
        }
    }


    //private static void divide_to_strings(List<String> nodes,String one,String two,String three,int index){
    public Solution divide_to_strings(List<String> nodes,List<String> divides,int index,int level,boolean jumper){
        level += 1;
        for (String s : divides){
            if (s.length() > 0 && s.charAt(0) == '0') return null;
        }

        if (nodes.size() == 0) return null;
        List<String> nodes_start = new ArrayList<>(nodes);
        List<String> divides_start = new ArrayList<>(divides);

        if (nodes_start.size() > 0){
            Solution jump = null;

            String node = nodes_start.remove(0);
            //preskoci
            if (jumper) jump = divide_to_strings(nodes_start,divides_start,index,level++,true);

            String replacor = divides_start.get(index);

            if (replacor.equals("") && node.equals("0")) return null;

            replacor += node;
            divides_start.set(index, replacor);


            Solution solution = control(divides_start.get(0),divides_start.get(1),divides_start.get(2));



            //bud ich da dokopy
            Solution unite = divide_to_strings(nodes_start,divides_start,index,level++,false);
            //alebo rozdeli
            Solution divide = index + 1 < 3 ? divide_to_strings(nodes_start,divides_start,index + 1 ,level++,false) : null ;
            Solution pom = choose_parrent1(unite,divide,jump,solution);
            return pom;
        }
        return null;
    }
    public boolean contains_all(Solution stones_father,Solution stones_son){
        for (String i : stones_son.stones){
            if (Collections.frequency(stones_father.stones,i) < Collections.frequency(stones_son.stones,i)) return false;
        }
        return true;
    }

    private int toInt(String val) {
        return val.equals("") ? -1 :Integer.parseInt(val);
    }

    public Solution choose_parrent1(Solution parent,Solution parent2 , Solution parent3 , Solution actual){
        // ak ma tri nadpriklady
        Solution out = null;
        List<Solution> parents = new ArrayList<Solution>(Arrays.asList(parent,parent2,parent3));
        if (Collections.frequency(parents,null) == 0){
            if (parent2.value >= parent.value && parent2.value >= parent3.value){
                parents.remove(parent2);
                out = parent2;
            }
            else if (parent3.value >= parent.value){
                parents.remove(parent3);
                out = parent3;
            }
            else {
                parents.remove(parent);
                out = parent;
            }

            for (Solution s : parents){
                if (contains_all(out,s)){
                    out.add_child(s);
                }
            }

        }
        //ak dvaja
        else if (Collections.frequency(parents,null) == 1){
            parents.remove(null);
            if (parents.get(0).value >= parents.get(1).value) {

                out = parents.remove(0); //odstrani lepsieho
            }
            else {
                out = parents.remove(1);
            }
            for (Solution s : parents){
                if (contains_all(out,s)){
                    out.add_child(s);
                }
            }
        }
        //ak iba jeden
        else if (Collections.frequency(parents,null) == 2){
            parents.removeAll(Collections.singleton(null));
            out = parents.get(0);
        }
        //ak ziaden
        else if (actual != null){
            out = actual;
        }

        if (actual != null && out != null){
            if (contains_all(out,actual)){
                out.add_child(actual);
            }
        }
        return out;
    }

    public Solution control(String one, String two,String three){

        int one_num = toInt(one);
        int two_num = toInt(two);
        int three_num = toInt(three);

        if (two_num == -1 && three_num == -1) return null;


        List<Solution> solution_values;
        if (three_num == -1){
            solution_values = check_unary_solution(one_num,two_num);
        }
        else {
            solution_values = check_binary_solution(one_num,two_num,three_num);
        }
        if (solution_values.size() > 0){
            Solution parent = solution_values.remove(0);

            for (Solution s : solution_values) parent.add_child(s);
            return parent;
        }
        return null;
    }

    public void check_best(Solution solution){
        boolean is_there = true;


        for (Solution s : solutions){
            if (s.priklad.equals(solution.priklad)) is_there = false;
        }

        if (solutions.size() < 10 && is_there){
            solutions.add(solution);
        }
        else if (solutions.get(0).value < solution.value && is_there){
            solutions.remove(0);
            solutions.add(solution);
        }
        Collections.sort(solutions);
    }


    public int count_digits(int in){
        int out = 0;
        while (in > 0){
            out += in % 10;
            in /= 10;
        }
        return out;
    }
    public List<String> make_stones_list(int one,int two,int three){
        String s = one + Integer.toString(two);
        if (three >= 0) s += Integer.toString(three);
        List<String> out = new ArrayList<>();
        for (char c : s.toCharArray()){
            out.add(Character.toString(c));
        }
        return out;
    }

    public List<Solution> check_unary_solution(int one_num,int two_num){
        List<Solution> out = new ArrayList<>();
        if (check_operator_in_play(Operators.POWER) && check_power_of(one_num,two_num)){
            out.add(new Solution(count_digits(one_num) + count_digits(two_num),one_num + " ** " + two_num,make_stones_list(one_num,two_num,-1)));
        }
        if (check_operator_in_play(Operators.ROOT) && check_root(one_num,two_num)){
            out.add(new Solution(count_digits(one_num) + count_digits(two_num),one_num + " /- " + two_num,make_stones_list(one_num,two_num,-1)));
        }
        return out;
    }
    public List<Solution> check_binary_solution(int one_num,int two_num,int three_num){
        List<Solution> out = new ArrayList<>();
        if (check_operator_in_play(Operators.PLUS) && check_plus(one_num,two_num,three_num)){
            out.add(new Solution(count_digits(one_num) + count_digits(two_num),one_num + " + " + two_num + " = " + three_num,make_stones_list(one_num,two_num,three_num)));
        }
        if (check_operator_in_play(Operators.MINUS) && check_minus(one_num,two_num,three_num)){
            out.add(new Solution(count_digits(one_num) + count_digits(two_num),one_num + " - " + two_num + " = " + three_num,make_stones_list(one_num,two_num,three_num)));
        }
        if (check_operator_in_play(Operators.MULTIPLY) && check_multiply(one_num,two_num,three_num)){
            out.add(new Solution(count_digits(one_num) + count_digits(two_num),one_num + " * " + two_num + " = " + three_num,make_stones_list(one_num,two_num,three_num)));
        }
        if (check_operator_in_play(Operators.DIVIDE) && check_divide(one_num,two_num,three_num)){
            out.add(new Solution(count_digits(one_num) + count_digits(two_num),one_num + " / " + two_num + " = " + three_num,make_stones_list(one_num,two_num,three_num)));
        }
        return  out;
    }

    public boolean check_operator_in_play(Operators o){
        return operations.contains(o);
    }

    public boolean check_power_of(int one_num,int two_num){
        return Math.pow(one_num,2) == two_num || Math.pow(one_num,3) == two_num;
    }
    public boolean check_root(int one_num,int two_num){
        return Math.sqrt(one_num) == two_num || Math.cbrt(one_num) == two_num;
    }
    public boolean check_plus(int one_num,int two_num,int three_num){
        return one_num + two_num == three_num;
    }
    public boolean check_minus(int one_num,int two_num,int three_num){
        return one_num - two_num == three_num;
    }
    public boolean check_multiply(int one_num,int two_num,int three_num){
        return one_num * two_num == three_num;
    }
    public boolean check_divide(int one_num,int two_num,int three_num){
        return (double)one_num / two_num == three_num;
    }
}
